writeaheadlog.go

```diff
@@ -225,30 +291,57 @@ func (w *WAL) recoverWal(data []byte) ([]Update, error) {
 	return updates, nil
 }
 
+// writeRecoveryState is a helper function that changes the recoveryState on disk
+func (w *WAL) writeRecoveryState(state uint16) error {
+	// Set the metadata to unclean again
+	recoveryState := make([]byte, 2)
+	binary.LittleEndian.PutUint16(recoveryState, state)
+	if _, err := w.logFile.WriteAt(recoveryState, 0); err != nil {
+		return err
+	}
+	return nil
+}
+
 // RecoveryComplete is called after a wal is recovered to signal that it is
 // safe to reset the wal
 func (w *WAL) RecoveryComplete() error {
-	// Marshal the pageStatusApplied
-	pageAppliedBytes := make([]byte, 8)
-	binary.LittleEndian.PutUint64(pageAppliedBytes, pageStatusApplied)
+	// Set the metadata to wipe
+	if err := w.writeRecoveryState(recoveryStateWipe); err != nil {
+		return err
+	}
 
-	// Get the length of the file.
-	stat, err := w.logFile.Stat()
-	if err != nil {
+	// Sync before we start swiping
+	if err := w.logFile.Sync(); err != nil {
 		return err
 	}
-	length := stat.Size()
 
-	// Set all pages to applied.
-	for offset := int64(pageSize + checksumSize); offset < length; offset += pageSize {
-		if _, err := w.logFile.WriteAt(pageAppliedBytes, offset); err != nil {
-			return err
-		}
+	// Simulate crash after recovery state was written
+	if w.deps.disrupt("RecoveryFail") {
+		return nil
 	}
+
+	// Wipe the wal
+	if err := w.wipeWAL(); err != nil {
+		return err
+	}
+
+	// Sync after swiping
+	if err := w.logFile.Sync(); err != nil {
+		return err
+	}
+
+	// Set the metadata to unclean again
+	if err := w.writeRecoveryState(recoveryStateUnclean); err != nil {
+		return err
+	}
+
 	// Sync to lock down the obliteration.
-	w.logFile.Sync()
+	if err := w.logFile.Sync(); err != nil {
+		return err
+	}
```

Comment:
Same thing here, these steps (including setting w.recoveryComplete to `true`) should be moved into `wipeWAL` because you need to perform them every time you wipe the wal.